src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/GroupCollection.cs

```diff
@@ -250,5 +250,69 @@ void IEnumerator.Reset()
 
             void IDisposable.Dispose() { }
         }
+
+        public bool TryGetValue(string key, out Group value)
+        {
+            if (ContainsKey(key))
+            {
+                value = this[key];
+                return true;
+            }
+            value = null;
+            return false;
+        }
+
+        public bool ContainsKey(string key)
+        {
+            return _match._regex.GroupNumberFromName(key) >= 0;
+        }
+
+        public IEnumerable<string> Keys
+        {
+            get
+            {
+                return _match._regex.GetGroupNames();
+            }
+        }
+
+        public IEnumerable<Group> Values
+        {
+            get
+            {
+                return this._groups;
```

Comment:
@ViktorHofer, are we ok returning this array directly?   The caller could cast back to `Group[]` and then mutate the array.  What would happen in that case, and do we care?

Reply:
Good catch. I haven't thought about that. The collection only lives on the caller's site => it is returned when calling Regex.Matches. We don't accept the GroupCollection as a parameter on any of the existing APIs, therefore I think we are fine with this side effect.

But I don't have a strong opinion here. I would also be fine with a copy returned.